Cherry-pick changes to PRNG algorithms

dhardy opened the pull request:
Notable changes:
-   `Rng::next_u64` and `Rng::fill_bytes` no longer have default implementations
-   `impls` module added (private; the idea is that this gets exported as `rand_core::impls` later)
-   Isaac code massively improved and performance improved by @pitdicker 
-   ChaCha code also improved to a lesser extent

Note that this is based on `core` (PR #208) which accounts for the changes to Cargo.toml and the README.

Also note: yes, I merged these 17 commits by hand (ugh) and ran the tests for each. Reviewing in detail from this PR will be horrible; it may be easier to review the original commits if you want to do that. But everything has already been well reviewed by @pitdicker and myself and there are a bunch of tests, so I'm not too worried.

dhardy commented:
Turns out I had to bump the Rust version to 0.17.0 to make Travis happy. If necessary we can tweak to make 1.15.0 happy (see https://github.com/dhardy/rand/issues/66), but unless someone actually requests this  I won't bother.

pitdicker commented:
I have read over the changes, and it looks like a lot of work backporting! It looks good, there is nothing I could comment on.

The change to make `IsaacWordRng` a wrapper instead of a 'use' is missing. But the only purpose of that change is for better documentation, and not that interesting for now.

dhardy commented:
Yes, I deliberately missed out `IsaacWordRng`; it's not exposed and I'm not sure it's useful.

dhardy commented:
Rebased (first commit moving `Issac64` to a new module merged to make diff easier to read)

dhardy commented:
Breaking changes (see also [here](https://github.com/rust-lang-nursery/rand/pull/211#discussion_r157379012)):

* `Rng::next_u64` and `Rng::fill_bytes` no longer have default implementations
* `StdRng` no longer implements `Copy`
* `new_unseeded` fns removed from ISAAC variants [replaced by `new_from_u64`; easy to provide a deprecated wrapper fn]
* `Isaac64Rng::next_u32` produces different results
* `Debug` implementations no longer show internals [unimportant]

Not sure at the moment whether to try to work around the above (not so easy) or make this wait until 0.5.

dhardy commented:
Ok, I reverted all the above breaking changes (except to `Debug`) so that this PR can land as 0.4.2.

The commits affecting `Isaac64Rng::next_u32` were removed entirely and will be re-added later.

dhardy commented on src/lib.rs:
```diff
@@ -334,20 +335,16 @@ pub trait Rand : Sized {
+    /// or even produce different results.
     fn next_u64(&mut self) -> u64 {
-        ((self.next_u32() as u64) << 32) | (self.next_u32() as u64)
+        impls::next_u64_via_u32(self)
```
This is also a breaking change (was BE, or possibly undefined order)

dhardy replied:
No, it turns out the old code also produced LE values (for me at least).

pitdicker replied:
Hmm, having that as a breaking change does not make me happy.

I double-checked your results, and if I add the test on the master branch it fails...

dhardy replied:
Thanks for picking that up. I forgot `IsaacRng` already has it's own impl of `next_u64`! Now I get BE results from the old code too. Ok, will revert.

dhardy commented:
Actually, this is more complicated: the `next_u64` impls for ChaCha and Xorshift also need to be removed, as do the new `fill_bytes` implementations. That means some of your optimisations are not applicable, unfortunately. So I'm not sure whether it's worth merging this in 0.4.x at all?

pitdicker commented:
Are you sure `fill_bytes` has changed? But I was just thinking the same thing, how much is left that can land? The code for `isaac::next_u64` can be adjusted by just swapping `x` and `y` though.

dhardy commented:
Well, there was change in output (I tried running the new `true_bytes` test with old code).

The problem here is that these changes affect output that was not previously tested, *and that we want to change anyway* (standardise to LE). I don't want to write new tests to confirm we didn't break output that was untested before and will soon be obsolete. So unless someone else does that, I think I'll just merge all [breaking changes](https://github.com/rust-lang-nursery/rand/pull/209#issuecomment-352438410) except for the removal of default implementations in `Rng` (which is an API change likely to hit many `Rng` impls) for 0.5.

pitdicker commented:
I was asking more because I was careful to preserve the order of `fill_bytes` in the "Fill `isaac64` backwards" commit. But that was after the switch to LE. Now that I see the old code, it was first combining two `u32` in a big-endian `u64`, and then using that to return bytes in what I think is little endian.

> I think I'll just merge all breaking changes except for the removal of default implementations in Rng (which is an API change likely to hit many Rng impls) for 0.5.

Seems like a good plan to me. As you say we want those changes in the end, so why go through all the hoops just to remain in the 0.4.* version.

I am not sure what you want to wait for to remove the default implementations in `Rng`. If there where many users of 0.4.* I would wait. And if there was a big upcoming version to collect these breaking changes. But that is not the release model we ended up with?

dhardy commented:
Updated (adjusted the "revert breaking changes" commit and squashed a couple of commits together to simplify). Due to breaking changes, this will land in 0.5.

This means that breaking changes are those [above](https://github.com/rust-lang-nursery/rand/pull/209#issuecomment-352438410) minus the removal of default implementations and of `new_unseeded`.

pitdicker commented:
🎉
